Userland/Libraries/LibCore/ResourceImplementation.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2023, Andrew Kaster <akaster@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/DirIterator.h>
+#include <LibCore/ResourceImplementation.h>
+#include <LibCore/ResourceImplementationFile.h>
+#include <LibCore/System.h>
+
+namespace Core {
+
+static OwnPtr<ResourceImplementation> s_the;
+
+void ResourceImplementation::install(OwnPtr<ResourceImplementation> the)
+{
+    s_the = move(the);
+}
+
+ResourceImplementation& ResourceImplementation::the()
+{
+    if (!s_the) {
+        install(make<ResourceImplementationFile>("/res"_string));
+    }
+    return *s_the;
+}
+
+NonnullOwnPtr<Resource> ResourceImplementation::make_resource(String full_path, NonnullOwnPtr<Core::MappedFile> file)
+{
+    return adopt_own(*new Resource(move(full_path), Resource::Scheme::Resource, move(file)));
+}
+
+NonnullOwnPtr<Resource> ResourceImplementation::make_resource(String full_path, ByteBuffer buffer)
+{
+    return adopt_own(*new Resource(move(full_path), Resource::Scheme::Resource, move(buffer)));
+}
+
+NonnullOwnPtr<Resource> ResourceImplementation::make_directory_resource(String full_path)
+{
+    return adopt_own(*new Resource(move(full_path), Resource::Scheme::Resource, Resource::DirectoryTag {}));
+}
+
+ErrorOr<NonnullOwnPtr<Resource>> ResourceImplementation::load_from_uri(StringView uri)
+{
+    StringView const file_scheme = "file://"sv;
+    StringView const resource_scheme = "resource://"sv;
+
+    if (uri.starts_with(resource_scheme))
+        return load_from_resource_scheme_uri(uri);
+
+    if (uri.starts_with(file_scheme)) {
+        auto path = uri.substring_view(file_scheme.length());
+        if (is_directory(path))
+            return adopt_own(*new Resource(TRY(String::from_utf8(path)), Resource::Scheme::File, Resource::DirectoryTag {}));
+        return adopt_own(*new Resource(TRY(String::from_utf8(path)), Resource::Scheme::File, TRY(MappedFile::map(path))));
+    }
+
+    dbgln("ResourceImplementation: Unknown scheme for {}", uri);
+    return Error::from_string_view("Invalid scheme"sv);
+}
+
+Vector<String> ResourceImplementation::child_names(Resource const& resource)
+{
+    if (!resource.is_directory())
+        return {};
+
+    if (resource.m_scheme == Resource::Scheme::Resource)
+        return child_names_for_resource_scheme(resource);
+
+    VERIFY(resource.m_scheme == Resource::Scheme::File);
+
+    Vector<String> children;
+    Core::DirIterator it(resource.filesystem_path().release_value().to_deprecated_string(), Core::DirIterator::SkipParentAndBaseDir);
```

Comment:
MUST() instead of release_value since it signifies more clearly that we're disregarding an unimportant OOM.

Reply:
I can't MUST() an optional